util/template/expression_template.go

```diff
@@ -17,12 +18,22 @@ func init() {
 }
 
 func expressionReplace(w io.Writer, expression string, env map[string]interface{}, allowUnresolved bool) (int, error) {
-	if _, ok := env["retries"]; !ok && hasRetries(expression) && allowUnresolved {
+	// The template is JSON-marshaled. This JSON-unmarshals the expression to undo any character escapes.
```

Comment:
we need to be very careful with expressions, I've seen many cases of patching specific issues, that just produce new issues

Reply:
Yep, there's a lot going on here.

I would _like_ to be more expressive with the change. Basically, it's "grab part of a JSON string, unescape it, evaluate it, and escape the result before re-inserting it in the string".

But the code is much more difficult to read.

Reply:
btw, it looks like example workflows can be invoked as e2e tests... if I can figure out that system, I'll add some example workflows which test at least these bugs.